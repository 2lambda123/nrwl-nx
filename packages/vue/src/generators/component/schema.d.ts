export interface Schema {
  name: string;
  project: string;
  skipTests?: boolean;
  directory?: string;
  export?: boolean;
  pascalCaseFiles?: boolean;
  pascalCaseDirectory?: boolean;
  routing?: boolean;
  js?: boolean;
  flat?: boolean;
  fileName?: string;
  inSourceTests?: boolean;
  skipFormat?: boolean;
  unitTestRunner?: 'jest' | 'vitest' | 'none';
}

export interface NormalizedSchema extends Schema {
  projectSourceRoot: string;
  fileName: string;
  className: string;
}
