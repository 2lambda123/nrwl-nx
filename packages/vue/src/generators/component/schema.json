{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "NxVueComponent",
  "title": "Create a Vue Component",
  "description": "Create a Vue Component for Nx.",
  "type": "object",
  "examples": [
    {
      "command": "nx g component my-component --project=mylib",
      "description": "Generate a component in the `mylib` library"
    },
    {
      "command": "nx g component my-component --project=mylib --classComponent",
      "description": "Generate a class component in the `mylib` library"
    }
  ],
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "alias": "p",
      "$default": {
        "$source": "projectName"
      },
      "x-prompt": "What is the name of the project for this component?",
      "x-priority": "important"
    },
    "name": {
      "type": "string",
      "description": "The name of the component.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the component?",
      "x-priority": "important"
    },
    "js": {
      "type": "boolean",
      "description": "Generate JavaScript files rather than TypeScript files.",
      "default": false
    },
    "skipTests": {
      "type": "boolean",
      "description": "When true, does not create `spec.ts` test files for the new component.",
      "default": false,
      "x-priority": "internal"
    },
    "directory": {
      "type": "string",
      "description": "Create the component under this directory (can be nested).",
      "alias": "dir",
      "x-priority": "important"
    },
    "flat": {
      "type": "boolean",
      "description": "Create component at the source root rather than its own directory.",
      "default": false
    },
    "export": {
      "type": "boolean",
      "description": "When true, the component is exported from the project `index.ts` (if it exists).",
      "alias": "e",
      "default": false,
      "x-prompt": "Should this component be exported in the project?"
    },
    "pascalCaseFiles": {
      "type": "boolean",
      "description": "Use pascal case component file name (e.g. `App.tsx`).",
      "alias": "P",
      "default": false
    },
    "pascalCaseDirectory": {
      "type": "boolean",
      "description": "Use pascal case directory name (e.g. `App/App.tsx`).",
      "alias": "R",
      "default": false
    },
    "routing": {
      "type": "boolean",
      "description": "Generate a library with routes."
    },
    "fileName": {
      "type": "string",
      "description": "Create a component with this file name."
    },
    "inSourceTests": {
      "type": "boolean",
      "default": false,
      "description": "When using Vitest, separate spec files will not be generated and instead will be included within the source files. Read more on the Vitest docs site: https://vitest.dev/guide/in-source.html"
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false,
      "x-priority": "internal"
    },
    "unitTestRunner": {
      "type": "string",
      "enum": ["vitest", "jest", "none"],
      "description": "Test runner to use for unit tests.",
      "x-prompt": "What unit test runner should be used?"
    }
  },
  "required": ["name", "project"]
}
