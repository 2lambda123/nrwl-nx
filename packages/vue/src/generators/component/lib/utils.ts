import {
  applyChangesToString,
  getProjects,
  joinPathFragments,
  logger,
  names,
  Tree,
} from '@nx/devkit';
import { NormalizedSchema, Schema } from '../schema';
import { ensureTypescript } from '@nx/js/src/utils/typescript/ensure-typescript';
import { addImport } from '../../../utils/ast-utils';

let tsModule: typeof import('typescript');

export async function normalizeOptions(
  host: Tree,
  options: Schema,
  isApp: boolean
): Promise<NormalizedSchema> {
  assertValidOptions(options);

  let { className, fileName } = names(options.name);
  const componentFileName =
    options.fileName ?? (options.pascalCaseFiles ? className : fileName);
  const project = getProjects(host).get(options.project);

  if (!project) {
    throw new Error(
      `Cannot find the ${options.project} project. Please double check the project name.`
    );
  }

  const { sourceRoot: projectSourceRoot, projectType } = project;

  className = getComponentClassName(
    className,
    isApp,
    options.project,
    options.directory
  );

  const directory = await getDirectory(options);

  if (options.export && projectType === 'application') {
    logger.warn(
      `The "--export" option should not be used with applications and will do nothing.`
    );
  }

  options.routing = options.routing ?? false;
  options.inSourceTests = options.inSourceTests ?? false;

  return {
    ...options,
    directory,
    className,
    fileName: componentFileName,
    projectSourceRoot,
  };
}

export function getComponentClassName(
  componentName: string,
  isApp: boolean,
  projectName: string,
  directory?: string
): string {
  const { className } = names(projectName);

  let prefix = isApp ? 'App' : className;

  if (directory?.length > 0) {
    prefix = directory
      .split('/')
      .map((segment) =>
        segment
          .split('-')
          .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
          .join('')
      )
      .join('');
  }

  return `${prefix}${componentName}`;
}

export function addExportsToBarrel(
  host: Tree,
  options: NormalizedSchema,
  isApp: boolean
) {
  if (!tsModule) {
    tsModule = ensureTypescript();
  }

  if (options.export && !isApp) {
    const indexFilePath = joinPathFragments(
      options.projectSourceRoot,
      options.js ? 'index.js' : 'index.ts'
    );
    const indexSource = host.read(indexFilePath, 'utf-8');
    if (indexSource !== null) {
      const indexSourceFile = tsModule.createSourceFile(
        indexFilePath,
        indexSource,
        tsModule.ScriptTarget.Latest,
        true
      );
      const changes = applyChangesToString(
        indexSource,
        addImport(
          indexSourceFile,
          `export { default as ${options.className} } from './${options.directory}/${options.fileName}.vue';`
        )
      );
      host.write(indexFilePath, changes);
    }
  }
}

export async function getDirectory(options: Schema) {
  if (options.directory) return options.directory;
  if (options.flat) return 'components';
  const { className, fileName } = names(options.name);
  const nestedDir = options.pascalCaseDirectory === true ? className : fileName;
  return joinPathFragments('components', nestedDir);
}

export function assertValidOptions(options: Schema) {
  const slashes = ['/', '\\'];
  slashes.forEach((s) => {
    if (options.name.indexOf(s) !== -1) {
      const [name, ...rest] = options.name.split(s).reverse();
      let suggestion = rest.map((x) => x.toLowerCase()).join(s);
      if (options.directory) {
        suggestion = `${options.directory}${s}${suggestion}`;
      }
      throw new Error(
        `Found "${s}" in the component name. Did you mean to use the --directory option (e.g. \`nx g c ${name} --directory ${suggestion}\`)?`
      );
    }
  });
}
