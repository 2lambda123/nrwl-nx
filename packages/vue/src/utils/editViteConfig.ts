/**
 *   const vuePluginImportLine =
    options.uiFramework === 'vue'
      ? `import vue from '@vitejs/plugin-vue';`
      : '';

  const vuePlugin = options.uiFramework === 'vue' ? `vue(),` : '';
 */
