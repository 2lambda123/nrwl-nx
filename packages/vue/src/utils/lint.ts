import {
  eslintPluginVueVersion,
  vueEslintConfigPrettierVersion,
  vueEslintConfigTypescriptVersion,
} from './versions';

export const extraEslintDependencies = {
  dependencies: {},
  devDependencies: {
    '@vue/eslint-config-prettier': vueEslintConfigPrettierVersion,
    '@vue/eslint-config-typescript': vueEslintConfigTypescriptVersion,
    'eslint-plugin-vue': eslintPluginVueVersion,
  },
};

export const extendVueEslintJson = (json: any) => {
  const { extends: pluginExtends, ...config } = json;

  return {
    extends: [
      'plugin:vue/vue3-essential',
      'eslint:recommended',
      '@vue/eslint-config-typescript',
      '@vue/eslint-config-prettier/skip-formatting',
      ...(pluginExtends || []),
    ],
    ...config,
  };
};
