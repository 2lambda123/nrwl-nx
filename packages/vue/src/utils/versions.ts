export const nxVersion = require('../../package.json').version;
export const vueVersion = '^3.3.4';
export const vueTscVersion = '^1.8.8';
export const vueRouterVersion = '^4.2.4';
export const vueTestUtilsVersion = '^2.4.1';
export const vueEslintConfigPrettierVersion = '^8.0.0';
export const vueEslintConfigTypescriptVersion = '^11.0.3';
export const vueTsconfigVersion = '^0.4.0';
export const tsconfigNode18Version = '^18.2.0';
export const eslintVersion = '^8.46.0';
export const eslintPluginVueVersion = '^9.16.1';
export const vitePluginVueVersion = '^4.3.1';
