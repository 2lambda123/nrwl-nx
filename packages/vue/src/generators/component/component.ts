import {
  formatFiles,
  generateFiles,
  GeneratorCallback,
  getProjects,
  joinPathFragments,
  runTasksInSerial,
  toJS,
  Tree,
} from '@nx/devkit';
import { NormalizedSchema, Schema } from './schema';
import { join } from 'path';
import { addExportsToBarrel, normalizeOptions } from './lib/utils';

export async function componentGenerator(host: Tree, schema: Schema) {
  const workspace = getProjects(host);
  const isApp = workspace.get(schema.project).projectType === 'application';

  const options = await normalizeOptions(host, schema, isApp);

  createComponentFiles(host, options);

  const tasks: GeneratorCallback[] = [];

  addExportsToBarrel(host, options, isApp);

  if (!options.skipFormat) {
    await formatFiles(host);
  }

  return runTasksInSerial(...tasks);
}

function createComponentFiles(host: Tree, options: NormalizedSchema) {
  const componentDir = joinPathFragments(
    options.projectSourceRoot,
    options.directory
  );

  generateFiles(host, join(__dirname, './files'), componentDir, {
    ...options,
    tmpl: '',
    unitTestRunner: options.unitTestRunner,
  });

  for (const c of host.listChanges()) {
    let deleteFile = false;

    if (
      (options.skipTests || options.inSourceTests) &&
      /.*spec.ts/.test(c.path)
    ) {
      deleteFile = true;
    }

    if (deleteFile) {
      host.delete(c.path);
    }
  }

  if (options.js) toJS(host);
}

export default componentGenerator;
