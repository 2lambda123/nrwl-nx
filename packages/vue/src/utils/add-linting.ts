import { Tree } from 'nx/src/generators/tree';
import { Linter, lintProjectGenerator } from '@nx/linter';
import { joinPathFragments } from 'nx/src/utils/path';
import { addDependenciesToPackageJson, runTasksInSerial } from '@nx/devkit';
import { NormalizedSchema } from '../generators/library/schema';
import { extraEslintDependencies } from './lint';
import {
  addExtendsToLintConfig,
  isEslintConfigSupported,
} from '@nx/linter/src/generators/utils/eslint-file';

export async function addLinting(
  host: Tree,
  options: {
    linter: Linter;
    name: string;
    projectRoot: string;
    unitTestRunner?: 'jest' | 'vitest' | 'none';
    setParserOptionsProject?: boolean;
    skipPackageJson?: boolean;
  },
  projectType: 'lib' | 'app'
) {
  if (options.linter === Linter.EsLint) {
    const lintTask = await lintProjectGenerator(host, {
      linter: options.linter,
      project: options.name,
      tsConfigPaths: [
        joinPathFragments(options.projectRoot, `tsconfig.${projectType}.json`),
      ],
      unitTestRunner: options.unitTestRunner,
      eslintFilePatterns: [`${options.projectRoot}/**/*.{ts,tsx,js,jsx,vue}`],
      skipFormat: true,
      setParserOptionsProject: options.setParserOptionsProject,
    });

    if (isEslintConfigSupported(host)) {
      addExtendsToLintConfig(host, options.projectRoot, [
        'plugin:vue/vue3-essential',
        'eslint:recommended',
        '@vue/eslint-config-typescript',
        '@vue/eslint-config-prettier/skip-formatting',
      ]);
    }

    let installTask = () => {};
    if (!options.skipPackageJson) {
      installTask = addDependenciesToPackageJson(
        host,
        extraEslintDependencies.dependencies,
        extraEslintDependencies.devDependencies
      );
    }

    return runTasksInSerial(lintTask, installTask);
  } else {
    return () => {};
  }
}
