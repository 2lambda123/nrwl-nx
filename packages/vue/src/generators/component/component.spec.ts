import { logger, readJson, Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { componentGenerator } from './component';
import { createLib } from '../../utils/test-utils';

describe('component', () => {
  let appTree: Tree;
  let projectName: string;

  beforeEach(async () => {
    projectName = 'my-lib';
    appTree = createTreeWithEmptyWorkspace();
    await createLib(appTree, projectName);
    jest.spyOn(logger, 'warn').mockImplementation(() => {});
    jest.spyOn(logger, 'debug').mockImplementation(() => {});
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should generate files with vitest', async () => {
    await componentGenerator(appTree, {
      name: 'hello',
      project: projectName,
      unitTestRunner: 'vitest',
    });

    expect(
      appTree.exists('my-lib/src/components/hello/hello.vue')
    ).toBeTruthy();
    expect(
      appTree.exists('my-lib/src/components/hello/__tests__/hello.spec.ts')
    ).toBeTruthy();

    expect(
      appTree.read('my-lib/src/components/hello/hello.vue', 'utf-8')
    ).toMatchSnapshot();
    expect(
      appTree.read(
        'my-lib/src/components/hello/__tests__/hello.spec.ts',
        'utf-8'
      )
    ).toMatchSnapshot();
  });

  it('should generate files with jest', async () => {
    await componentGenerator(appTree, {
      name: 'hello',
      project: projectName,
      unitTestRunner: 'jest',
    });

    expect(
      appTree.read('my-lib/src/components/hello/hello.vue', 'utf-8')
    ).toMatchSnapshot();
    expect(
      appTree.read(
        'my-lib/src/components/hello/__tests__/hello.spec.ts',
        'utf-8'
      )
    ).toMatchSnapshot();
  });

  // we don't have app generator yet
  xit('should generate files for an app', async () => {
    await componentGenerator(appTree, {
      name: 'hello',
      project: 'my-app',
      unitTestRunner: 'vitest',
    });

    expect(
      appTree.exists('my-app/src/components/hello/hello.tsx')
    ).toBeTruthy();
    expect(
      appTree.exists('my-app/src/components/hello/hello.spec.ts')
    ).toBeTruthy();
    expect(
      appTree.exists('my-app/src/components/hello/hello.module.css')
    ).toBeTruthy();
  });

  describe('--export', () => {
    it('should add to index.ts barrel', async () => {
      await componentGenerator(appTree, {
        name: 'hello',
        project: projectName,
        export: true,
      });
      expect(appTree.read('my-lib/src/index.ts', 'utf-8')).toMatchSnapshot();
    });

    // no app generator yet
    xit('should not export from an app', async () => {
      await componentGenerator(appTree, {
        name: 'hello',
        project: 'my-app',
        export: true,
      });

      expect(appTree.read('my-app/src/index.ts', 'utf-8')).toMatchSnapshot();
    });
  });

  describe('--pascalCaseFiles', () => {
    it('should generate component files with upper case names', async () => {
      await componentGenerator(appTree, {
        name: 'hello',
        project: projectName,
        pascalCaseFiles: true,
      });
      expect(
        appTree.exists('my-lib/src/components/hello/Hello.vue')
      ).toBeTruthy();
      expect(
        appTree.exists('my-lib/src/components/hello/__tests__/Hello.spec.ts')
      ).toBeTruthy();
    });
  });

  describe('--pascalCaseDirectory', () => {
    it('should generate component files with pascal case directories', async () => {
      await componentGenerator(appTree, {
        name: 'hello-world',
        project: projectName,
        pascalCaseFiles: true,
        pascalCaseDirectory: true,
      });
      expect(
        appTree.exists('my-lib/src/components/HelloWorld/HelloWorld.vue')
      ).toBeTruthy();
      expect(
        appTree.exists(
          'my-lib/src/components/HelloWorld/__tests__/HelloWorld.spec.ts'
        )
      ).toBeTruthy();
    });
  });

  // TODO: figure out routing
  xdescribe('--routing', () => {
    it('should add routes to the component', async () => {
      await componentGenerator(appTree, {
        name: 'hello',
        project: projectName,
        routing: true,
      });

      const content = appTree
        .read('my-lib/src/components/hello/hello.tsx')
        .toString();
      expect(content).toContain('react-router-dom');
      expect(content).toMatch(/<Route\s*path="\/"/);
      expect(content).toMatch(/<Link\s*to="\/"/);

      const packageJSON = readJson(appTree, 'package.json');
      expect(packageJSON.dependencies['react-router-dom']).toBeDefined();
    });
  });

  describe('--directory', () => {
    it('should create component under the directory', async () => {
      await componentGenerator(appTree, {
        name: 'hello',
        project: projectName,
        directory: 'components',
      });

      expect(appTree.exists('/my-lib/src/components/hello/hello.vue'));
    });

    it('should create with nested directories', async () => {
      await componentGenerator(appTree, {
        name: 'helloWorld',
        project: projectName,
        directory: 'lib/foo',
      });

      expect(
        appTree.exists('/my-lib/src/components/foo/hello-world/hello-world.vue')
      );
    });
  });

  describe('--flat', () => {
    it('should create in project directory rather than in its own folder', async () => {
      await componentGenerator(appTree, {
        name: 'hello',
        project: projectName,
        flat: true,
      });

      expect(appTree.exists('/my-lib/src/components/hello.vue'));
    });
    it('should work with custom directory path', async () => {
      await componentGenerator(appTree, {
        name: 'hello',
        project: projectName,
        flat: true,
        directory: 'components',
      });

      expect(appTree.exists('/my-lib/src/components/hello.vue'));
    });
  });
});
