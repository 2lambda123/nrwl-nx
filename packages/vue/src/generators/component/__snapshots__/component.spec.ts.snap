// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component --export should add to index.ts barrel 1`] = `
"export { default as MyLibHello } from './components/hello/hello.vue';
"
`;

exports[`component --export should not export from an app 1`] = `null`;

exports[`component should generate files with jest 1`] = `
"<script setup lang="ts">
defineProps<{}>();
</script>

<template>
  <p>Welcome to MyLibHello!</p>
</template>

<style scoped></style>
"
`;

exports[`component should generate files with jest 2`] = `
"import { mount } from '@vue/test-utils';
import MyLibHello from './hello.vue';

describe('MyLibHello', () => {
  it('renders properly', () => {
    const wrapper = mount(MyLibHello, {});
    expect(wrapper.text()).toContain('Welcome to MyLibHello');
  });
});
"
`;

exports[`component should generate files with vitest 1`] = `
"<script setup lang="ts">
defineProps<{}>();
</script>

<template>
  <p>Welcome to MyLibHello!</p>
</template>

<style scoped></style>
"
`;

exports[`component should generate files with vitest 2`] = `
"import { describe, it, expect } from 'vitest';

import { mount } from '@vue/test-utils';
import MyLibHello from './hello.vue';

describe('MyLibHello', () => {
  it('renders properly', () => {
    const wrapper = mount(MyLibHello, {});
    expect(wrapper.text()).toContain('Welcome to MyLibHello');
  });
});
"
`;
